### PR TITLE
[mle] fix some corner cases when trying to attach to a better partition

### PR DESCRIPTION
**Pull request by librasungirl**
Thanks for @zhaxia's observations during stress tests on Cert_5_5_08 when not removing tmp/ directory and his help on validating these fixes.

### CODE REVIEW

File: src/core/thread/mle_router.cpp [hunk 1]
```diff
@@ -519,7 +519,15 @@ bool MleRouter::HandleAdvertiseTimer(void)
         return false;
     }
 
-    SendAdvertisement();
+    // Suppress MLE Advertisements when trying to attach to a better partition.
+    //
+    // Without this suppression, a device may send an MLE Advertisement before receiving the MLE Child ID Response.
+    // The candidate parent then removes the attaching device because the Source Address TLV includes an RLOC16 that
+    // indicates a Router role (i.e. a Child ID equal to zero).
+    if (mParentRequestState == kParentIdle)
```
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: Move this check inside `SendAdvertisement()`?

Comment (librasungirl) [reply src/core/thread/mle_router.cpp]: Updated. Thanks.

File: src/core/thread/mle.cpp [hunk 2]
```diff
@@ -1647,11 +1647,23 @@ otError Mle::SendChildIdRequest(void)
...
+        }
+        else
+        {
+            // Invalidate stale parent state.
+            //
+            // Parent state is not normally invalidated after becoming a Router/Leader (see #1875).  When trying to
+            // attach to a better partition, invalidating old parent state (especially when in kStateRestored) ensures
+            // that GetNeighbor() returns mParentCandidate when processing the Child ID Response.
+            mParent.SetState(Neighbor::kStateInvalid);
+            otPlatSettingsDelete(&GetInstance(), Settings::kKeyParentInfo, -1);
```
Comment (abtink) [inline src/core/thread/mle.cpp]: Is there any reason/advantage to remove the old parent from the settings here? Checking the existing code, we don't do this from `BecomeDetached()`. I think the idea was that the settings entry would be overwritten once we get attached to new parent.

File: src/core/thread/mle.cpp [hunk 3]
```diff
@@ -1647,11 +1647,23 @@ otError Mle::SendChildIdRequest(void)
...
+            ExitNow(error = OT_ERROR_ALREADY);
+        }
+        else
+        {
+            // Invalidate stale parent state.
+            //
+            // Parent state is not normally invalidated after becoming a Router/Leader (see #1875).  When trying to
+            // attach to a better partition, invalidating old parent state (especially when in kStateRestored) ensures
+            // that GetNeighbor() returns mParentCandidate when processing the Child ID Response.
+            mParent.SetState(Neighbor::kStateInvalid);
```
Comment (abtink) [inline src/core/thread/mle.cpp]: Nice find!

File: src/core/thread/mle.cpp (hunk 2 above)
Comment (jwhui) [reply src/core/thread/mle.cpp]: Overwriting the new parent after attaching is too late in this case.  As noted in the comment, `GetNeighbor()` will return `mParent` rather than `mParentCandidate` and causes the attach to fail.

Comment (librasungirl) [reply src/core/thread/mle.cpp]: It is not overwritten, but just written into a new logic block. And the parent information which is last written would be retrieved in `otPlatSettingsGet()`. 

The advantage may be not so obvious, but here marking the outdated information with deleted flag in flash would help to improve the efficiency of 
1) getting parent info from flash (false for this `if` to avoid more flash op) https://github.com/openthread/openthread/blob/master/examples/platforms/utils/settings.cpp#L363)
2) finding the valid information when swapping is required (false for this `if` to avoid extra loop and flash op)
 (https://github.com/openthread/openthread/blob/master/examples/platforms/utils/settings.cpp#L175)
 
It do have some subtle advantages when the device may experience multiple attach processes after long-term run.

Thoughts?

Comment (abtink) [reply src/core/thread/mle.cpp]: Thanks. Sounds good to me.
